README.md
# Boostrap continuous token deployment repository

This project is based on [Aragon One Court presale project](https://github.com/aragonone/court-presale) and provides a similar fashion presale and bonding curve based AMM to link two tokens.

## Important changes and future work

- This contracts were decoupled from Aragon's ENS integration. This means, some contracts are manually deployed and initialized (requiring the system contracts to extend from `UnsafeAragonApp` instead).
- Provide a sample `IContinuousToken` implementation for those who are not willing to provide their own.

## How to integrate with this project

Among other things that are detailed in the project config, there is an important requirement that this project has: the bonded / continuous token must comply with the [IContinuousToken](./contracts/IContinuousToken.sol) interface. This way, both the Presale and the MarketMaker (the bonding curve based market making contract) will require to `mint` and `burn` the continuous token.

Due to the fact that the continuous token might require some special rules defined in they project they belong to, it's important to define and deploy such token using the separate project. To do so, please implement the before mentioned interface so everything could work as expected.

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites

What things you need to install the software and how to install them

```bash
# Install nvm
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
# Install proper node version
nvm use
```

Create `.env` file (you can base on [`.env.example`](./.env.example))

### Installing

A step by step series of examples that tell you how to get a development env running

Say what the step will be

```bash
# Install the dependencies
yarn
```

### Generate Types

In order to get contract types you can generate those typings when compiling

```bash
yarn compile
```

## Running the tests

```bash
yarn test
```

### Testing with Waffle

Tests using Waffle are written with Mocha alongside with Chai. 

Is recommended to use Gherkin as a language to describe the test cases

```
describe("Feature: Greeter", () => {
  describe("Scenario: Should return the new greeting once it's changed", () => {
    let greeter: Greeter;
    it("GIVEN a deployed Greeter contract", async () => {
      const factory = await ethers.getContractFactory("Greeter");
      greeter = <Greeter>await factory.deploy("Hello, world!");
      expect(await greeter.greet()).to.equal("Hello, world!");
    });
    it("WHEN greeting message changed", async () => {
      await greeter.setGreeting("Hola, mundo!");
    });
    it("THEN greet returns new greeting message", async () => {
      expect(await greeter.greet()).to.equal("Hola, mundo!");
    });
  });
});
```

We are requiring Chai which is an assertions library. These asserting functions are called "matchers", and the ones we're using here actually come from Waffle.

For more information we suggest reading waffle testing documentation [here](https://hardhat.org/guides/waffle-testing.html#testing).

### Ethereum Security Toolbox

To check the code statically you can use the Ethereum Security Toolbox made by Trail of Bits.

##### Default option

You could just run the default checking executing

`yarn security-default`

##### Flexibile option

Or if you want more flexibility, first execute the command 

`yarn security`

and once inside the docker container run 

```
solc-select 0.4.24
cd project
```

so that you can use the tools there installed.

## Tasks & Deployment

### Custom tasks

There were added some custom `hardhat` tasks to help with the token manipulation and deployment mostly for testing or debugging purposes.

```
npx hardhat help

initialize                      initialize bonding curve contracts and set permissions
close-presale                   closes the presale and let's people to start trading
contribute                     	buys (during the presale period) some bonded tokens and sends them to the recipient
get-state                       returns presale current state
mint-collateral                 mints some collateral tokens (SOV) and sends them to the recipient address
open-presale                    starts the presale
update-presale-date             Testing command that updates the mocked presale date to a specific value so state can be changed
open-buy-order                  open a buy order of bonded tokens after presale period
claim-buy-order                 claim a buy order of bonded tokens
open-sell-order                 open a sell order of bonded tokens after presale period
claim-sell-order                claim a sell order of bonded tokens
print-system-info               prints system useful data to montior the presale and the bonding curve status
emergency-remove-from-reserve   Recovers the funds from the reserve
```

For example, if the `MockedBalanceRedirect` presale was deployed, the following commands can be executed in order to open, contribute and then close it:

```bash
# Initialize bonding curve contracts and set permissions
npx hardhat initialize --network rskTestnetMocked
# Move forward in time to make the sale open
npx hardhat update-presale-date --span "2 days" --network rskTestnetMocked
# Mint some SOV (collateral)
npx hardhat mint-collateral --network rskTestnetMocked --recipient "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B" --amount "10000000000000000000000"
# Buy some Zero (bonded)
npx hardhat contribute --network rskTestnetMocked --recipient "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B" --amount "10000000000000000000000"
# Close the presale
npx hardhat update-presale-date --span "6 weeks" --network rskTestnetMocked
npx hardhat close-presale --network rskTestnetMocked
# Open buy order (buy bonded tokens sending collateral tokens) 
npx hardhat open-buy-order --network rskTestnetMocked --amount "1000"
# Claim buy order (use buy order's batch id)
npx hardhat claim-buy-order --network rskTestnetMocked --batch "229110"
# Open sell order (buy collateral tokens sending bonded tokens) 
npx hardhat open-sell-order --network rskTestnetMocked --amount "1000"
# Claim sell order (use sell order's batch id)
npx hardhat claim-sell-order --network rskTestnetMocked --batch "229120"
```

### Deployment

Use `hardhat.config.ts` to set deployment parameters for each network. Parameters description can be found in such file.

```bash
yarn deploy:dev         # launches a `buidlervm` instance and deploys the contracts    
yarn deploy:dev:fixture      # generates some tokens and assigns them to some hardcoded accounts for testing purposes
yarn deploy:rskdev      # deploys the contracts to a local RSK node
yarn devchain:start     # launches an RSK regtest node. Requires Docker to be installed.
yarn deploy:rskTestnetMocked # deploys to RSK tesnet with mocked Presale Contract so the state can be tweaked
yarn initialize:rskTestnetMocked  #initialize deployed contracts and set permissions
```

### Funds recovery

In case funds needs to be recovered from the reserve, the following steps:

1. Disable opening buy and sell orders (done by the governance account)
2. Granting funds retirements permission to an account (done by the governance account)
3. Execute the transfer (by the account that has such permissions, it could be the governance itself)

There is a task that will let execute `(2)` and `(3)` steps but a complete example can be found in [this test](test/presale-interaction.ts).

## Built With

* [Aragon One Court presale project](https://github.com/aragonone/court-presale)
* [Atix Labs](https://atixlabs.com) template
* [Hardhat](https://hardhat.org/) - Task runner

## Contributing

Please read [CONTRIBUTING.md](./CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests.

## Versioning

We use [SemVer](http://semver.org/) and [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) for versioning. For the versions available, see the [tags on this repository](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/tags).

To create a new release execute the script

`yarn release`

## License

This project is licensed under the MIT License - see the [LICENSE.md](https://github.com/IQAndreas/markdown-licenses) file for details


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/compare/v1.0.2...v1.1.0) (2021-11-08)


### Features

* emergency-remove-from-reserve task and documentation ([12aa1b7](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/12aa1b710f6fc691f53a61e6b6763ef74c8fb0e4))
* improved print-system-info task to show more data about the presale ([9ee2caa](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/9ee2caa268bbff3c470c20870eea6b9228561b88))
* mynt mainnet deployment addresses ([4dd449b](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/4dd449bf7e03bf282c257815e206899e7c21e254))
* mynt production deployment configs ([be35d54](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/be35d54d0c6f2f203d5d3bde5d31e4b71a4c5c54))
* setup of Mynt RSK testnet deployment ([0bfd628](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/0bfd628770c584ab6ad8693abdfc641c7773ac94))
* updated MyntToken address after deployment ([e216281](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/e216281ad26aeea90271534a0b49509dfd434543))


### Bug Fixes

* beneficiaryPercent was being rounded to 15 instead of being 15.75 ([e3222e5](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/e3222e551d667d27364c59ef137401dffebd15b0))
* fixed print-system-info task so it can work with already deployed tokens ([a09a41b](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/a09a41bcf25e41836cca97b3897b3124a34e4e1c))

### [1.0.2](https://sovryn.github.com/swamp-thing-sovryn/zero-continuous-token-presale/compare/v1.0.1...v1.0.2) (2021-11-03)

### [1.0.1](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/compare/v1.0.0...v1.0.1) (2021-11-03)


### Bug Fixes

* added some logs and made it work on some steps were it was failing ([1dc9aff](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/1dc9aff9c2bddf6de153783a540a53da78888f1c))

## [1.0.0](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/compare/v0.0.1-alpha.1...v1.0.0) (2021-11-03)


### Features

* print system info command ([b14586b](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/b14586ba42a82b79f228cd2c8c25a56372f67501))
* **tests:** add revoke buy and sell order permissions and transfer reserve funds tests ([9c8d0d7](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/9c8d0d735937b47ff371629b94cc2155104da0b3))

### [0.0.1-alpha.1](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/compare/v0.0.1-alpha.0...v0.0.1-alpha.1) (2021-10-28)


### Features

* configure ACL in a single transaction ([e39eeff](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/e39eeff2c8224f94e47ae9b84a7f146cbd2f562c))

### 0.0.1-alpha.0 (2021-10-25)


### Features

* add Agent and Vault init ([a6e1889](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/a6e1889257dcf66e8d7a62e73dfdf0f305715bee))
* add buy and sell order tasks ([e1de514](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/e1de5143a46584966398d047dbd835ed53bec050))
* add continuous token and remove token manager ([572a74a](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/572a74a1cf08fd997e2870db2a6a4b22ba41d878))
* add initialize script ([#4](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/issues/4)) ([93bcdf8](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/93bcdf89221b7dd428baae43bcf80920c022b585))
* added ACL configs - WIP ([f8c0881](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/f8c0881a73979223274df476b3a7b6275c769240))
* added dao deployment - still wip ([a4d7da0](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/a4d7da03d5e0bdcea136b9459776fca8efa64dd2))
* added waffleMatchers-deploy-fixtures ([a6225f1](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/a6225f1fb28585a5e42e73f44c569a52efad94b4))
* deploy tokens only local ([952fa28](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/952fa28277f11520b9b494f0b05eb45885ad3238))
* deploy with flags ([14a0f54](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/14a0f545aa4eb1a6a64f4ea4787c3c167e73b54a))
* deployment can be configured to just deploy a single mocked token ([eb303e3](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/eb303e3eadbe0d2c5c6044f8dc1d5ab2b2146821))
* deploys in separate files ([8b7b73f](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/8b7b73f5eaf6792cd7832cb52c13034610052df7))
* rsk deployment & helpers ([40c3603](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/40c36039c51c019ffb1cf8914673b1c5b98c4016))
* set deployment parameters in config ([7a5a25e](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/7a5a25e09b48cb77962988ef01732ec8efd0e4bc))
* set deployment parameters in hre ([509e5ae](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/509e5ae6f2937a2f4f58f965ff43dc023eeb1473))
* **test:**  init bonding curve tests ([3091119](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/3091119b7ce17d877333a5199b62090f966a1e02))
* **test:**  presale  tests ([630c949](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/630c94921419af1c34989cc1ffcc6980cf81d989))
* **test:** market maker tests ([ddccdac](https://github.com/swamp-thing-sovryn/zero-continuous-token-presale/commit/ddccdac2bfc1a8e649c610b68d1ae17c130be441))

hardhat.config.ts
import "@atixlabs/hardhat-time-n-mine";
import "@nomiclabs/hardhat-ethers";
import "@nomiclabs/hardhat-solhint";
import "@nomiclabs/hardhat-waffle";
import "@openzeppelin/hardhat-upgrades";
import "@typechain/hardhat";
// This is done to have the new matchers from waffle,
// because despite the note in https://hardhat.org/guides/waffle-testing.html?#adapting-the-tests
// the changeEtherBalance is not added because its a newer version
import chai from "chai";
import { config as dotenvConfig } from "dotenv";
import { solidity } from "ethereum-waffle";
import "hardhat-contract-sizer";
import "hardhat-deploy";
import "hardhat-docgen";
import "hardhat-gas-reporter";
import "hardhat-preprocessor";
import { removeConsoleLog } from "hardhat-preprocessor";
import "hardhat-prettier";
import { HardhatUserConfig } from "hardhat/types";
import { resolve } from "path";
import { BigNumber, Wallet } from "ethers";
import "solidity-coverage";
import "./scripts/custom-tasks";
import ms from "ms";
import { addMilliseconds } from "date-fns";

chai.use(solidity);

dotenvConfig({ path: resolve(__dirname, "./.env") });

const chainIds = {
  ganache: 1337,
  hardhat: 31337,
  rskTestnetMocked: 31,
  rskMainnet: 30,
};

const PPM = BigNumber.from(1e6);
const PCT_BASE = BigNumber.from((1e18).toString());
const DAYS = 24 * 3600;

const startInAnHourFromNow = () => {
  const delta = ms("1 hour");
  const newDate = addMilliseconds(new Date(), delta);
  return BigNumber.from(newDate.getTime()).div(1000);
};

// Ensure that we have all the environment variables we need.
let mnemonic: string;
if (!process.env.MNEMONIC) {
  throw new Error("Please set your MNEMONIC in a .env file");
} else {
  mnemonic = process.env.MNEMONIC;
}

const getPrivateKey = () =>
  process.env.DEPLOYER_PRIVATE_KEY ? [process.env.DEPLOYER_PRIVATE_KEY] : [Wallet.createRandom().privateKey];

type Parameters = {
  // the date (unixtime, ms) upon which that presale is to be open [ignored if 0]
  startDate: BigNumber;
  // the percentage of the raised funds that will be sent to beneficiary address during presale period
  mintingBeneficiaryPCT: BigNumber;
  // the amount of time, in ms, that the presale will last
  presalePeriod: number;
  // the exchangeRate [= 1/price] at which [bonded] tokens are to be purchased for that presale [in PPM]
  presaleEchangeRate: BigNumber;
  // the reserve ratio to be used for that collateral token [in PPM]
  reserveRatio: BigNumber;
  // the amount of blocks a batch will contain
  batchBlock: number;
  // the price slippage below which each batch is to be kept for that collateral token [in PCT_BASE]
  slippage: BigNumber;
  // the fee to be deducted from buy orders [in PCT_BASE]
  buyFee: BigNumber;
  // the fee to be deducted from sell orders [in PCT_BASE]
  selFee: BigNumber;
  // the address of the collateral token, only necessary. If not provided, a mock token will be deployed
  collateralTokenAddress?: string;
  // the address of the bonded token, only necessary. if not provided a mock token will be deployed.
  bondedTokenAddress?: string;
  // the address of the governance, permissions will be transfer to this address after deployment. If not provided, permissions remains to deployer address
  governanceAddress?: string;
  // the address of the beneficiary, fees will be transfer to this address.
  beneficiaryAddress: string;
};
declare module "hardhat/types/config" {
  export interface HardhatNetworkUserConfig {
    // If true it will deploy a mocked presale contract version that allow update date to close presale (for testing purposes)
    mockPresale: boolean;
    parameters: Parameters;
  }

  export interface HardhatNetworkConfig {
    mockPresale: boolean;
    parameters: Parameters;
  }
}

const config: HardhatUserConfig = {
  defaultNetwork: "hardhat",
  namedAccounts: {
    deployer: 0,
    otherUser: 1,
  },
  networks: {
    hardhat: {
      mockPresale: true,
      parameters: {
        startDate: BigNumber.from(new Date().getTime()).div(1000).add(DAYS),
        mintingBeneficiaryPCT: PPM.mul(35).div(100),
        presalePeriod: 14 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        beneficiaryAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
        governanceAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
      },
      accounts: {
        mnemonic,
      },
      chainId: chainIds.hardhat,
    },
    rskdev: {
      mockPresale: false,
      parameters: {
        startDate: BigNumber.from(new Date().getTime()).div(1000).add(1000),
        mintingBeneficiaryPCT: PPM.mul(35).div(100),
        presalePeriod: 14 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        beneficiaryAddress: "0xB0D1D7fad89CfC28394b0B1AB51d24c432170f5A",
      },
      url: "http://localhost:4444",
      // regtest default prefunded account
      from: "0xcd2a3d9f938e13cd947ec05abc7fe734df8dd826",
      gasMultiplier: 1.25,
    },
    rskTestnetMocked: {
      mockPresale: true,
      parameters: {
        startDate: BigNumber.from(new Date().getTime()).div(1000).add(1000),
        mintingBeneficiaryPCT: PPM.mul(35).div(100),
        presalePeriod: 2 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        beneficiaryAddress: "",
        governanceAddress: "",
      },
      url: "https://public-node.testnet.rsk.co",
      accounts: getPrivateKey(),
      chainId: chainIds.rskTestnetMocked,
    },
    rskTestnetMockedWithSOV: {
      mockPresale: false,
      parameters: {
        startDate: startInAnHourFromNow(),
        mintingBeneficiaryPCT: PPM.mul(35).div(100),
        presalePeriod: 1 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        collateralTokenAddress: "0x6a9A07972D07e58F0daf5122d11E069288A375fb",
        beneficiaryAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
        governanceAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
      },
      url: "https://testnet.sovryn.app/rpc",
      accounts: getPrivateKey(),
      chainId: chainIds.rskTestnetMocked,
    },
    myntRSKTestnet: {
      mockPresale: false,
      parameters: {
        startDate: startInAnHourFromNow(),
        mintingBeneficiaryPCT: PPM.mul(BigNumber.from("1575")).div(100).div(100),
        presalePeriod: 1 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        collateralTokenAddress: "0x6a9A07972D07e58F0daf5122d11E069288A375fb",
        bondedTokenAddress: "0x139483e22575826183F5b56dd242f8f2C1AEf327",
        beneficiaryAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
        governanceAddress: "0x4D1A9fD1E1e67E83Ffe72Bdd769088d689993E4B",
      },
      url: "https://testnet.sovryn.app/rpc",
      accounts: getPrivateKey(),
      chainId: chainIds.rskTestnetMocked,
    },
    myntRSKMainnet: {
      mockPresale: false,
      parameters: {
        startDate: BigNumber.from("1636383600"), // "2021-10-11T15:00:00.000Z"
        mintingBeneficiaryPCT: PPM.mul(BigNumber.from("1575")).div(100).div(100),
        presalePeriod: 7 * DAYS,
        presaleEchangeRate: PPM.mul(10000).div(100),
        reserveRatio: PPM.mul(40).div(100),
        batchBlock: 10,
        slippage: PCT_BASE.mul(3).div(100),
        buyFee: BigNumber.from(0),
        selFee: PCT_BASE.mul(3).div(1000),
        collateralTokenAddress: "0xefc78fc7d48b64958315949279ba181c2114abbd",
        bondedTokenAddress: "0x2e6B1d146064613E8f521Eb3c6e65070af964EbB",
        beneficiaryAddress: "0x924f5ad34698Fd20c90Fe5D5A8A0abd3b42dc711",
        governanceAddress: "0x924f5ad34698Fd20c90Fe5D5A8A0abd3b42dc711",
      },
      url: "https://mainnet.sovryn.app/rpc",
      accounts: getPrivateKey(),
      chainId: chainIds.rskMainnet,
      timeout: 20000 * 100
    },
  },
  paths: {
    artifacts: "./artifacts",
    cache: "./cache",
    sources: "./contracts",
    tests: "./test",
  },
  solidity: {
    version: "0.4.24",
    settings: {
      // https://hardhat.org/hardhat-network/#solidity-optimizer-support
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  typechain: {
    outDir: "typechain",
    target: "ethers-v5",
    alwaysGenerateOverloads: false,
  },
  gasReporter: {
    enabled: process.env.REPORT_GAS ? true : false,
    currency: "USD",
    gasPrice: 21,
  },
  preprocess: {
    eachLine: removeConsoleLog(hre => !["hardhat", "localhost"].includes(hre.network.name)),
  },
  docgen: {
    path: "./docs",
    clear: true,
    runOnCompile: false,
  },
  contractSizer: {
    alphaSort: true,
    runOnCompile: true,
    disambiguatePaths: false,
  },
  mocha: {
    timeout: 60000,
  },
};

config.networks;
export default config;
